plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -218,38 +256,33 @@ public MetadataEntry getMetadataEntry(TableSnapshot tableSnapshot, ConnectorSess
     public List<AddFileEntry> getActiveFiles(TableSnapshot tableSnapshot, ConnectorSession session)
     {
         try {
-            CacheKey cacheKey = new CacheKey(tableSnapshot.getTable(), tableSnapshot.getTableLocation());
-            DeltaLakeDataFileCacheEntry cachedTable = activeDataFileCache.get(cacheKey, () -> {
-                List<AddFileEntry> activeFiles = loadActiveFiles(tableSnapshot, session);
-                return new DeltaLakeDataFileCacheEntry(tableSnapshot.getVersion(), activeFiles);
-            });
-            if (cachedTable.getVersion() > tableSnapshot.getVersion()) {
-                log.warn("Query run with outdated Transaction Log Snapshot, retrieved stale table entries for table: %s and query %s", tableSnapshot.getTable(), session.getQueryId());
-                return loadActiveFiles(tableSnapshot, session);
-            }
-            if (cachedTable.getVersion() < tableSnapshot.getVersion()) {
-                DeltaLakeDataFileCacheEntry updatedCacheEntry;
-                try {
+            TableVersion tableVersion = new TableVersion(new TableLocation(tableSnapshot.getTable(), tableSnapshot.getTableLocation()), tableSnapshot.getVersion());
+
+            DeltaLakeDataFileCacheEntry cacheEntry = activeDataFileCache.get(tableVersion, () -> {
+                DeltaLakeDataFileCacheEntry oldCached = activeDataFileCache.asMap().keySet().stream()
```

Comment:
Could the key set here grow very large? With a high cache TTL, frequent version updates of the tables, and frequent queries querying multiple tables it seems plausible. Though, decreasing TTL could solve it.

Reply:
that's why this is a cache. it is even weighted by size of its entries

Reply:
Still - you are iterating over keys here. How many keys do we expect in real case scenario?

Reply:
Added `activeDataFileCacheIndex`, but I am not sure I like the solution.
Basing the logic on cache removal listener seems fragile.
PTAL

Reply:
I will extract the indexing complexity to https://github.com/trinodb/trino/pull/19163 and revert to the simpler version we had here.

Reply:
Maybe we could use a [ConcurrentSkipListMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentSkipListMap.html)?

Reply:
not sure how this helps?
anyway, let's have this discussion under  https://github.com/trinodb/trino/pull/19163

Reply:
... unless it's considered required for merging this

Reply:
It should not be a blocker for merging this.

I was wondering if the `floorEntry` method of `ConcurrentSkipListMap` could be used instead of a filter or a secondary index.